File: Algorithms/C++/0202/0202.cpp
class Solution {
public:
    bool isHappy(int n) {
        if (n <= 0) {
            return false;
        }
        set<int> s;
        while (n != 1) {
            int sum = 0;
            while (n) {
                int r = n % 10;
                sum += r * r;
                n /= 10;
            }
            n = sum;
            if (s.find(n) != s.end()) {
                return false;
            } else {
                s.insert(n);
            }
        }
        return true;
    }
};

File: Algorithms/Java/0202/0202.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def getNext(n: int) -> int:
            sum = 0
            while n:
                n, r =  divmod(n, 10)
                sum += r ** 2
            return sum
        m = getNext(n)
        while m != 1 and m != n:
            n = getNext(n)
            m = getNext(getNext(m))
        return m == 1
